Catchable items, playable basketball

slarticodefast opened the pull request:
## About the PR
Adds a component that makes it possible to catch items that are thrown at you and adds it to balls.
Adds a clumsiness interaction making you take a small amount of damage from failing to catch something.

## Why / Balance
Now you can actually play basketball on the basketball shuttle.

## Technical details
Added a simple `CatchableComponent` and corresponding system.

On a thrown item collision check if the target has free hands, is in combat mode (if required), passes an optional whitelist, and decide if you catch the item with a certain chance. To make the randomness not mispredict we set the seed depending on the current game tick and the entity's net id. Ideally this will be later replaced by https://github.com/space-wizards/RobustToolbox/pull/5849

I applied the same randomness prediction trick to `ClumsySystem`, which fixes https://github.com/space-wizards/space-station-14/issues/38538 and resolves https://github.com/space-wizards/space-station-14/issues/38369

Added the `CatchableComponent` to the basketball, football and beach ball and adjust their fixtures to be better.

## Media
![example](https://github.com/user-attachments/assets/e094dd9a-fce4-47d7-938d-35ed22285948)

![1](https://github.com/user-attachments/assets/70d80ede-f43c-4e07-958d-462afd0aa981)

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
Moved and renamed clumsiness related Loc strings.

**Changelog**
:cl:
- add: You can now catch basketballs, footballs and beach balls that are thrown at you!

ArtisticRoomba commented:
![zedorfska-ss13](https://github.com/user-attachments/assets/aeecd59f-3fc3-483c-816e-4988558f4f8f)

keronshb commented:
Thank DOG. This was missing from combat too, so the possibilities with this now are gonna be great.

slarticodefast commented:
> Thank DOG. This was missing from combat too, so the possibilities with this now are gonna be great.

Yeah, I was thinking about some mechanics for that. Maybe add a bool datafield to make it only possible to catch that item while in combat mode? How does this work in ss13?

But we can easily expand this later, for now: basketball.

ScarKy0 commented:
> Maybe add a bool datafield to make it only possible to catch that item while in combat mode?

Image of a secoff just straight up parrying a soap thrown by the clown.
Please. This sounds so awesome.

slarticodefast commented:
> Image of a secoff just straight up parrying a soap thrown by the clown.

New combat strategy: Throw two basketballs at someone to block their hands :godo:

keronshb commented:
> Yeah, I was thinking about some mechanics for that. Maybe add a bool datafield to make it only possible to catch that item while in combat mode? How does this work in ss13?

On cit it worked in throwmode. Which we don't have but we can delagate it to combat mode.

DogZeroX commented:
Need to make hot potato use this

Krunklehorn commented:
Awww hyeah, can't wait to do another station-wide pitching contest.

![vlc_J0B9R6oRwp](https://github.com/user-attachments/assets/e2e027cb-07ab-4ae4-b2fc-a23a4ad90487)

RedBookcase commented:
At long last

![knife-toss](https://github.com/user-attachments/assets/6841bb63-1da1-4523-8f1f-43b808236929)

beck-thompson commented:
Maybe add a clumsy interaction?

keronshb commented:
> Maybe add a clumsy interaction?

Nah, being able to catch things as a Clown is part of the job.

TeenSarlacc commented:
> > Image of a secoff just straight up parrying a soap thrown by the clown.
> 
> New combat strategy: Throw two basketballs at someone to block their hands :godo:

throw a glued minibomb at the captain :trollface:

Nukesthestation commented:
wait holy shit thats actually kind of op, just throw a glued minibomb at anyone and gib them

> throw a glued minibomb at the captain :trollface:

TeenSarlacc commented:
> wait holy shit thats actually kind of op, just throw a glued minibomb at anyone and gib them
> 
> > throw a glued minibomb at the captain :trollface:
> 
> 

just make sure you don't get it stuck in your own hand :godo:

slarticodefast commented:
I added a clumsiness interaction (and fixed a prediction bug in ClumsySystem along the way).
The `CatchableComponent` now also has a `RequiresCombatMode` bool and an optional whitelist for the catcher. This should allow usecases like martial arts, where someone with a certain component would be able to catch grenades for example.

beck-thompson commented on Content.Shared/Clumsy/ClumsySystem.cs:
```diff
@@ -7,29 +7,34 @@
     [Dependency] private readonly IGameTiming _timing = default!;
     [Dependency] private readonly IConfigurationManager _cfg = default!;
+    [Dependency] private readonly INetManager _net = default!;
+
```
Extra white space

slarticodefast commented:
I fixed the loc strings, but it turns out the hypospray and climbing are both completely mispredicting due to randomness. You can check the first one with the solution editor looking at your bloodstream - it does not do what the predicted popups are saying. Defib and guns are fine because those are not predicted at the moment.
I'll look into it, so don't merge for now.

slarticodefast commented:
> I'd honestly prefer if this was a blacklist rather than a whitelist, being able to toss items to eachother would be an awesome convenience, and not to mention catching and throwing back grenades would be so sick

It can just be whitelisted to `ItemComponent` in that case. But not all items should be caught automatically, otherwise spears, arrows etc would not work anymore because you just catch them instead of getting hit.
Also it might be a little problematic if you can just block someone's hands by throwing anything at them. I'm planning to do some testing with this to figure out what the best UX would be, for example making it combat mode only. So for now, only basketballs.

slarticodefast commented:
Ok, fixed. Turns out `HashCode.Combine` initializes with a random number, so it will create different results on the server and client. So instead I used a very simple hash function I added to the random helpers. This will all be replaced with `RandomPredicted` at a later point. I also fixed the same error in `SharedBloodStreamSystem`.

beck-thompson reviewed (approved):
Looks good!
